toolchain/check/impl_lookup.cpp

```diff
@@ -347,6 +348,109 @@ static auto LookupImplWitnessInSelfFacetValue(
   return EvalImplLookupResult::MakeNonFinal();
 }
 
+// Substitutes witnesess in place of `LookupImplWitness` queries into `.Self`,
+// when the witness is for the same interface as the one `.Self` is referring
+// to.
+//
+// This allows access to the `FacetType` and its constraints from the witness,
+// and allows `ImplWitnessAccess` instructions to be immediately resolved to a
+// more specific value when possible.
+class SubstWitnessesCallbacks : public SubstInstCallbacks {
+ public:
+  // `context` must not be null.
+  explicit SubstWitnessesCallbacks(
+      Context* context, SemIR::LocId loc_id,
+      llvm::ArrayRef<SemIR::SpecificInterface> interfaces,
+      llvm::ArrayRef<SemIR::InstId> witness_inst_ids)
+      : SubstInstCallbacks(context),
+        loc_id_(loc_id),
+        interfaces_(interfaces),
+        witness_inst_ids_(witness_inst_ids) {}
+
+  auto Subst(SemIR::InstId& inst_id) const -> bool override {
+    // `FacetType` can be concrete even when it has rewrite constraints that
+    // have a symbolic dependency on `.Self`. See use of
+    // `GetConstantValueIgnoringPeriodSelf` in eval. So in order to recurse into
+    // `FacetType` we must not return `true` when it's concrete.
+    if (context().constant_values().Get(inst_id).is_concrete() &&
+        !context().insts().Is<SemIR::FacetType>(inst_id)) {
+      return true;
+    }
+
+    // As an optimization: FacetValue is already resolved, it does not have
+    // any substitutable .Self references.
+    if (context().insts().Is<SemIR::FacetValue>(inst_id)) {
+      return true;
+    }
+
+    if (!context().insts().Is<SemIR::LookupImplWitness>(inst_id)) {
+      return false;
+    }
```

Comment:
I wonder if instead of special-casing `LookupImplWitness`es naming `.Self`, we could instead special-case the `.Self` `BindSymbolicName` instruction itself, regardless of which instruction it's used from. That might allow you to do a little less unwrapping here, and also seems like it would be important for things like `X where .T = .Self`.

Reply:
Oh, no we do that a test for that. I believe we're good, as these are also `ImplWitnessAccess`es, which is what we actually want to cause to resolve to a value.